File: libs/shape/src/main/kotlin/mono/shape/shape/Text.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.shape

import kotlin.math.max
import mono.graphics.geo.Point
import mono.graphics.geo.Rect
import mono.graphics.geo.Size
import mono.shape.extra.ShapeExtra
import mono.shape.extra.TextExtra
import mono.shape.serialization.AbstractSerializableShape
import mono.shape.serialization.SerializableText

/**
 * A text shape which contains a bound and a text.
 *
 * TODO: Resize bound by text
 */
class Text(
    rect: Rect,
    id: String? = null,
    parentId: String? = null,
    isTextEditable: Boolean = true
) : AbstractShape(id = id, parentId = parentId) {
    private var userSettingSize: Size = Size.ZERO
        set(value) {
            field = value.takeIf { it.width >= 3 && it.height >= 3 } ?: Size.ZERO
        }

    // Text can be auto resized by text
    override var bound: Rect = rect

    val contentBound: Rect
        get() = if (extra.boundExtra.isBorderEnabled) {
            Rect.byLTWH(bound.left + 1, bound.top + 1, bound.width - 2, bound.height - 2)
        } else {
            bound
        }

    var text: String = ""
        private set
    var isTextEditable: Boolean = isTextEditable
        private set

    var isTextEditing: Boolean = false
        private set

    override var extra: TextExtra = TextExtra.withDefault()
        private set

    var renderableText: RenderableText = RenderableText.EMPTY
        private set

    constructor(
        startPoint: Point,
        endPoint: Point,
        id: String? = null,
        parentId: String? = null,
        isTextEditable: Boolean
    ) : this(
        Rect.byLTRB(startPoint.left, startPoint.top, endPoint.left, endPoint.top),
        id = id,
        parentId = parentId,
        isTextEditable = isTextEditable
    )

    internal constructor(serializableText: SerializableText, parentId: String?) : this(
        serializableText.bound,
        id = serializableText.id,
        parentId = parentId,
        isTextEditable = serializableText.isTextEditable
    ) {
        extra = TextExtra(serializableText.extra)
        setText(serializableText.text)
        versionCode = serializableText.versionCode
    }

    init {
        userSettingSize = rect.size
        updateRenderableText()
    }

    override fun toSerializableShape(isIdIncluded: Boolean): AbstractSerializableShape =
        SerializableText(
            id.takeIf { isIdIncluded },
            versionCode,
            bound,
            text,
            extra.toSerializableExtra(),
            isTextEditable
        )

    override fun setBound(newBound: Rect) = update {
        val isUpdated = bound != newBound
        userSettingSize = newBound.size
        bound = newBound

        updateRenderableText()

        isUpdated
    }

    fun setText(newText: String) = update {
        val isTextChanged = newText != text
        text = newText
        updateRenderableText()
        isTextChanged
    }

    fun setTextEditingMode(isEditing: Boolean) = update {
        val isUpdated = isTextEditing != isEditing
        isTextEditing = isEditing
        isUpdated
    }

    override fun setExtra(newExtra: ShapeExtra) {
        check(newExtra is TextExtra) {
            "New extra is not a TextExtra (${newExtra::class})"
        }
        if (newExtra == extra) {
            return
        }
        update {
            extra = newExtra
            updateRenderableText()

            true
        }
    }

    private fun updateRenderableText() {
        val maxRowCharCount =
            if (extra.hasBorder()) bound.width - 2 else bound.width
        if (text != renderableText.text ||
            maxRowCharCount != renderableText.maxRowCharCount
        ) {
            renderableText = RenderableText(text, max(maxRowCharCount, 1))
        }
    }

    fun makeTextEditable() {
        if (isTextEditable) {
            return
        }
        update {
            isTextEditable = true
            true
        }
    }

    /**
     * A class to generate renderable text.
     */
    class RenderableText(val text: String, val maxRowCharCount: Int) {
        private var renderableText: List<String>? = null
        fun getRenderableText(): List<String> {
            val nonNullRenderableText = renderableText ?: createRenderableText()
            renderableText = nonNullRenderableText
            return nonNullRenderableText
        }

        private fun createRenderableText(): List<String> =
            if (maxRowCharCount == 1) {
                text.map { it.toString() }
            } else {
                standardizeLines(text.split("\n"))
            }

        private fun standardizeLines(lines: List<String>): List<String> = lines
            .flatMap { line ->
                val adjustedLines = mutableListOf(StringBuilder())
                for (word in line.toStandardizedWords(maxRowCharCount)) {
                    val lastLine = adjustedLines.last()
                    val space = if (lastLine.isNotEmpty()) " " else ""
                    val newLineLength = lastLine.length + space.length + word.length
                    if (newLineLength <= maxRowCharCount) {
                        lastLine.append(space).append(word)
                    } else {
                        adjustedLines.add(StringBuilder(word))
                    }
                }
                adjustedLines
            }
            .map { it.toString() }

        private fun String.toStandardizedWords(maxCharCount: Int): List<String> =
            split(" ")
                .flatMap { word ->
                    if (word.length <= maxCharCount) listOf(word) else word.chunked(maxCharCount)
                }

        companion object {
            val EMPTY = RenderableText("", 0)
        }
    }
}

File: monosketch-svelte/src/lib/mono/shape/shape/text.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { Point } from "$libs/graphics-geo/point";
import { Rect } from "$libs/graphics-geo/rect";
import { Size } from "$libs/graphics-geo/size";
import { type ShapeExtra, TextExtra } from "$mono/shape/extra/shape-extra";
import { AbstractSerializableShape, SerializableText } from "$mono/shape/serialization/serializable-shape";
import { AbstractShape } from "$mono/shape/shape/abstract-shape";

/**
 * A text shape which contains a bound and a text.
 */
class Text extends AbstractShape {
    private userSettingSize: Size = Size.ZERO;

    // Text can be auto resized by text
    private boundInner: Rect;

    private textInner: string = "";
    private isTextEditableInner: boolean;
    private isTextEditingInner: boolean = false;
    private renderableTextInner: RenderableText = RenderableText.EMPTY;

    constructor(rect: Rect, id: string | null = null, parentId: string | null = null, isTextEditable: boolean = true) {
        super(id, parentId);
        this.boundInner = rect;
        this.isTextEditableInner = isTextEditable;
        this.userSettingSize = rect.size;
        this.updateRenderableText();
    }

    static fromPoints(startPoint: Point, endPoint: Point, id: string | null = null, parentId: string | null = null, isTextEditable: boolean): Text {
        const rect = Rect.byLTRB(startPoint.left, startPoint.top, endPoint.left, endPoint.top);
        return new Text(rect, id, parentId, isTextEditable);
    }

    static fromSerializable(serializableText: SerializableText, parentId: string | null): Text {
        const text = new Text(serializableText.bound, serializableText.actualId, parentId, serializableText.isTextEditable);
        text.extraInner = TextExtra.fromSerializable(serializableText.extra);
        text.setText(serializableText.text);
        text.versionCode = serializableText.versionCode;
        return text;
    }

    get text(): string {
        return this.textInner;
    }

    setText(newText: string): void {
        this.update(() => {
            const isTextChanged = newText !== this.textInner;
            this.textInner = newText;
            this.updateRenderableText();
            return isTextChanged;
        });
    }

    get isTextEditable(): boolean {
        return this.isTextEditableInner;
    }

    setTextEditingMode(isEditing: boolean): void {
        this.update(() => {
            const isUpdated = this.isTextEditingInner !== isEditing;
            this.isTextEditingInner = isEditing;
            return isUpdated;
        });
    }

    get contentBound(): Rect {
        return this.extra.boundExtra.isBorderEnabled
            ? Rect.byLTWH(this.bound.left + 1, this.bound.top + 1, this.bound.width - 2, this.bound.height - 2)
            : this.bound;
    }

    get bound(): Rect {
        return this.boundInner;
    }

    set bound(newBound: Rect) {
        this.update(() => {
            const isUpdated = this.boundInner !== newBound;
            this.userSettingSize = newBound.size;
            this.boundInner = newBound;
            this.updateRenderableText();
            return isUpdated;
        });
    }

    get extra(): TextExtra {
        return this.extraInner as TextExtra;
    }

    setExtra(newExtra: ShapeExtra): void {
        if (!(newExtra instanceof TextExtra)) {
            throw new Error(`New extra is not a TextExtra (${newExtra.constructor.name})`);
        }
        if (newExtra.equals(this.extraInner)) {
            return;
        }
        this.update(() => {
            this.extraInner = newExtra;
            this.updateRenderableText();
            return true;
        });
    }

    toSerializableShape(isIdIncluded: boolean): AbstractSerializableShape {
        return new SerializableText(
            this.id,
            !isIdIncluded,
            this.versionCode,
            this.bound,
            this.text,
            this.extra.toSerializableExtra(),
            this.isTextEditable,
        );
    }

    private updateRenderableText(): void {
        const maxRowCharCount = this.extra.hasBorder() ? this.bound.width - 2 : this.bound.width;
        if (this.text !== this.renderableTextInner.text || maxRowCharCount !== this.renderableTextInner.maxRowCharCount) {
            this.renderableTextInner = new RenderableText(this.text, Math.max(maxRowCharCount, 1));
        }
    }

    makeTextEditable(): void {
        if (this.isTextEditableInner) {
            return;
        }
        this.update(() => {
            this.isTextEditableInner = true;
            return true;
        });
    }
}

/**
 * A class to generate renderable text.
 */
class RenderableText {
    private renderableText: string[] | null = null;

    constructor(public text: string, public maxRowCharCount: number) {
    }

    getRenderableText(): string[] {
        if (!this.renderableText) {
            this.renderableText = this.createRenderableText();
        }
        return this.renderableText;
    }

    private createRenderableText(): string[] {
        if (this.maxRowCharCount === 1) {
            return this.text.split('');
        } else {
            return this.standardizeLines(this.text.split('\n'));
        }
    }

    private standardizeLines(lines: string[]): string[] {
        return lines.flatMap(line => {
            const adjustedLines: string[] = [];
            let currentLine = '';
            for (const word of this.toStandardizedWords(line, this.maxRowCharCount)) {
                const space = currentLine ? ' ' : '';
                const newLineLength = currentLine.length + space.length + word.length;
                if (newLineLength <= this.maxRowCharCount) {
                    currentLine += space + word;
                } else {
                    adjustedLines.push(currentLine);
                    currentLine = word;
                }
            }
            adjustedLines.push(currentLine);
            return adjustedLines;
        });
    }

    private toStandardizedWords(line: string, maxCharCount: number): string[] {
        return line.split(' ').flatMap(word => {
            if (word.length <= maxCharCount) {
                return [word];
            } else {
                return word.match(new RegExp(`.{1,${maxCharCount}}`, 'g')) || [];
            }
        });
    }

    static EMPTY = new RenderableText('', 0);
}
